### PR TITLE
Filter panel

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Source/Ui/OverlayPanel.cpp [hunk 1]
```diff
@@ -171,25 +159,35 @@ void OverlayPanel::position() {
...
-      layout()->setContentsMargins(DROP_SHADOW_WIDTH(), 0,
-        DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT());
-      update();
-      move({x, parent_bottom_left.y() + scale_height(1)});
+    if(x < screen_geometry.left()) {
+      x = screen_geometry.left() - DROP_SHADOW_WIDTH();
+    } else if(x + panel_size.width() > screen_geometry.right()) {
+      x = screen_geometry.right() - panel_size.width() + DROP_SHADOW_WIDTH();
     }
+    auto rect = [=] () -> QRect {
```
Comment (spireken) [inline Applications/Spire/Source/Ui/OverlayPanel.cpp:175]: You can capture by reference here.

File: Applications/Spire/Source/Ui/OverlayPanel.cpp [hunk 2]
```diff
@@ -171,25 +159,35 @@ void OverlayPanel::position() {
...
+          panel_size.height() + 1), panel_size};
+      } else {
+        auto margins = QMargins(DROP_SHADOW_WIDTH(), 0, DROP_SHADOW_WIDTH(),
+          DROP_SHADOW_HEIGHT());
+        layout()->setContentsMargins(margins);
+        return {QPoint(x, parent_bottom_left.y() + 1), panel_size};
+      }
+    }();
+    setGeometry(rect);
+    update();
```
Comment (spireken) [inline Applications/Spire/Source/Ui/OverlayPanel.cpp:191]: Does setting the geometry not cause a repaint?

File: Applications/Spire/Source/Ui/FilterPanel.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  return *m_component;
+}
+
+connection FilterPanel::connect_reset_signal(
+    const ResetSignal::slot_type& slot) const {
+  return m_reset_signal.connect(slot);
+}
+
+bool FilterPanel::event(QEvent* event) {
+  if(event->type() == QEvent::ShowToParent) {
```
Comment (spireken) [inline Applications/Spire/Source/Ui/FilterPanel.cpp:66]: Can you explain how this works?

**Review by spireken — CHANGES_REQUESTED** [review]
Looks good.

File: Applications/Spire/Source/Ui/FilterPanel.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    QWidget* parent)
+    : QWidget(parent),
+      m_component(component) {
+  auto layout = new QVBoxLayout(this);
+  layout->setSpacing(0);
+  layout->setContentsMargins(scale_width(MARGIN_SIZE),
+    scale_height(MARGIN_SIZE), scale_width(MARGIN_SIZE),
+    scale_height(MARGIN_SIZE));
+  auto header = new TextBox(title);
+  header->setEnabled(false);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilterPanel.cpp:37]: The ```TextBox``` is not disabled in the spec.

File: Applications/Spire/Source/Ui/FilterPanel.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    font.setWeight(QFont::Medium);
+    font.setPixelSize(scale_width(12));
+    style.get(Any()).
+      set(border_size(0)).
+      set(text_style(font, QColor::fromRgb(0x80, 0x80, 0x80)));
+    return style;
+  }
+}
+
+FilterPanel::FilterPanel(const QString& title, QWidget* component,
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilterPanel.cpp:27]: Use move semantics in the constructor.

File: Applications/Spire/Source/Ui/OverlayPanel.cpp [hunk 6]
```diff
@@ -171,25 +159,35 @@ void OverlayPanel::position() {
     auto parent_geometry = parentWidget()->rect();
     auto parent_bottom_left = parentWidget()->mapToGlobal(
       parent_geometry.bottomLeft());
-    auto screen_geometry =
-      get_current_screen(parent_bottom_left)->availableGeometry();
+    auto screen_geometry = parentWidget()->screen()->availableGeometry();
+    auto panel_size = [=] {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OverlayPanel.cpp:163]: Capture by reference.

File: Applications/Spire/Source/Ui/OverlayPanel.cpp [hunk 7]
```diff
@@ -171,25 +159,35 @@ void OverlayPanel::position() {
...
+      if(layout()->contentsMargins() == DROP_SHADOW_MARGINS()) {
+        return size() - QSize(0, DROP_SHADOW_HEIGHT());
+      }
+      return size();
+    }();
     auto x = parent_bottom_left.x() - DROP_SHADOW_WIDTH();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OverlayPanel.cpp:169]: Use IIFE.

**Review by kranar — CHANGES_REQUESTED** [review]
Looks good overall.

File: Applications/Spire/Source/Ui/OverlayPanel.cpp (hunk 2 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/OverlayPanel.cpp:191]: No, the setGeometry doesn't update the panel when the panel's size and layout changes. When the panel moves with its parent and may be placed above or below its parent, the layout can't be updated in real time.

File: Applications/Spire/Source/Ui/FilterPanel.cpp (hunk 3 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/FilterPanel.cpp:66]: The FilterPanel is the child component of the OverlayPanel, and when the user shows the FilterPanel, it should tell its parent (OverlayPanel) to show up.

**Review by spireken — APPROVED** [review]

**Review by kranar — APPROVED** [review]
This was excellent.

**Pull request merged**